akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -471,8 +483,29 @@ private[projection] class R2dbcOffsetStore(
 
       val offsetInserts = dao.insertTimestampOffsetInTx(conn, filteredRecords)
 
-      offsetInserts.map { _ =>
-        if (state.compareAndSet(oldState, evictedNewState))
+      val adoptable =
+        if (oldState.foreignLatestSortedByTimestamp.isEmpty || filteredRecords.isEmpty) Vector.empty
+        else {
+          val latestTimestamp = filteredRecords.maxBy(_.timestamp).timestamp
+          oldState.foreignLatestSortedByTimestamp
+            .takeWhile(_.record.timestamp.compareTo(latestTimestamp) <= 0)
+            .map { adoptable =>
+              LatestBySlice(adoptable.record.slice, adoptable.record.pid, adoptable.record.seqNr)
+            }
+        }
+
+      val adoptedNewState =
+        if (adoptable.nonEmpty)
+          evictedNewState.copy(foreignLatestSortedByTimestamp =
+            oldState.foreignLatestSortedByTimestamp.drop(adoptable.size))
+        else evictedNewState
+
+      val adoptedOffsets =
+        if (adoptable.nonEmpty) offsetInserts.flatMap(_ => dao.adoptTimestampOffsets(adoptable))
```

Comment:
Currently after the offset inserts. Could be a separate maintenance task like deletion, and then putting the adoptable offsets, or the current timestamp, somewhere to be picked up.

Reply:
Updated to a separate scheduled task in https://github.com/akka/akka-projection/pull/1218/commits/2f13b7abe95a3b6197784ced4facaad6f1e88bc4